src/core/ConsoleVariable.h

```diff
@@ -4,15 +4,73 @@
 #include <nlohmann/json.hpp>
 #include <stdint.h>
 #include <memory>
-#include <map>
 #include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <variant>
 
 namespace Ship {
 typedef enum class ConsoleVariableType { Integer, Float, String, Color, Color24 } ConsoleVariableType;
 typedef union CVarValue {
 
 } CVarValue;
 
+class CVarInterface {
+  public:
+    using VauleType = std::variant<int32_t, float, std::string, Color_RGBA8, Color_RGB8>;
+
+    CVarInterface() = default;
+    virtual ~CVarInterface() = default;
+
+    virtual VauleType Get() const = 0;
+
+    virtual void Set(const VauleType&) = 0;
+
+    virtual bool ShouldSave() = 0;
+};
+
+template<class StorageType>
+class CVarDefaulted : public CVarInterface {
```

Comment:
This stores a CVar that has a default value. This is so we know to save this CVar only when it is not its default. This is to answer a question, "When asked to save, when should we save a CVar?" This is a question because these changes require a user to register all cvars in advance, so we don't necessarily want to save _all_ of them if many of them are unchanged since they were registered.

This was a struggle to me, since we are somewhat messy about things as it is. There are two possibilities:

- Save when they are not the default
- Save if they have ever been set (not just initialized).

I opted to do the first option since it allows use to change defaults if we ever decide a different default than what we had in a previous version is the best option for users. I am open to other opinions, however.